src/sardana/macroserver/macros/discretepm.py

```diff
@@ -0,0 +1,124 @@
+from sardana.macroserver.macro import Macro, Type
+import json
+import math
+
+
+class DiscretePseudoMotorConfiguration(dict):
+    def __init__(self, pseudo_obj, macro):
+        self.pseudo = pseudo_obj
+        self.macro = macro
+        _physical_motor_name = self.pseudo.physical_elements[0]
+        self.motor = macro.getMoveable(_physical_motor_name)
+        conf = self.get_configuration()
+        super(DiscretePseudoMotorConfiguration, self).__init__(conf)
+
+    def get_configuration(self):
+        return json.loads(self.pseudo.read_attribute('configuration').value)
+
+    def has_calibration(self):
+        return all(['set' in self[x].keys() for x in self.keys()])
+
+    def add_point(self, label, pos, dmin, dmax, set):
+        point = dict()
+        point['pos'] = int(pos)
+        # Calculate point calibration if required
+        if self.has_calibration():
+            # Set to current physical position if no value supplied as argument
+            if math.isinf(set):
+                point['set'] = self.motor.position
+            else:
+                point['set'] = float(set)
+            # If point exists, we use current min, max values
+            if label in self.keys() and math.isinf(dmin) and math.isinf(dmax):
+                p = self[label]
+                min_pos = point['set'] - abs(p['set'] - p['min'])
+                max_pos = point['set'] + abs(p['set'] - p['max'])
+            # else, new point has new calibration,
+            else:
+                min_pos = point['set'] - dmin
+                max_pos = point['set'] + dmax
+
+            point['min'] = min_pos
+            point['max'] = max_pos
+
+        self[label] = point
+        self._update()
+
+    def remove_point(self, label):
+        try:
+            self.pop(label)
+            self._update()
+        except Exception as e:
+            self.macro.error('Cannot remove label {0}\n{1}'.format(label, e))
+
+    def _update(self):
+        try:
+            self.pseudo.write_attribute('configuration', json.dumps(self))
```

Comment:
`self.pseudo` is a Taurus device. Tt is better to use taurus API to write attribute.

Reply:
Done